### PR TITLE
feat: Count API 

### PR DESCRIPTION
**Pull request by jainsahab**
Add support for querying the Datastore backend to just get the count (i.e. the size of the result set) for a query, instead of getting the entire result set.

### CODE REVIEW

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+public class AggregationResults implements Iterable<AggregationResult> {
+
+  private final List<AggregationResult> aggregationResults;
+  private Timestamp readTime;
+
+  public AggregationResults(List<AggregationResult> aggregationResults,
```
Comment (gforgurups) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java:29]: As its public constructor we may have to check for null condition for aggregationResults else all method may fail with unexpected behavior.

**Review by gforgurups — CHANGES_REQUESTED** [review]

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQueryProtoPreparer.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import com.google.cloud.datastore.StructuredQuery.OrderBy;
+import com.google.cloud.datastore.execution.request.ProtoPreparer;
+import com.google.datastore.v1.Query;
+import com.google.protobuf.Int32Value;
+
+public class StructuredQueryProtoPreparer implements ProtoPreparer<StructuredQuery<?>, Query> {
+
+  @Override
+  public Query prepare(StructuredQuery<?> query) {
+    com.google.datastore.v1.Query.Builder queryPb = com.google.datastore.v1.Query.newBuilder();
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQueryProtoPreparer.java:27]: Do you mean `query` (method parameter) ?

[the remark this answers is not in the document]
Comment (gforgurups) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQueryProtoPreparer.java:27]: Yes that's right. In addition, is com.google.datastore.v1.Query.newBuilder() is considered safe enough not to return null?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  public static class Builder {
+
+    private String namespace;
+    private Mode mode;
+    private final List<Aggregation> aggregations;
+    private StructuredQuery<?> nestedStructuredQuery;
+    private GqlQuery<?> nestedGqlQuery;
+
+    public Builder() {
+      this.aggregations = new ArrayList<>();
```
Comment (gforgurups) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java:76]: Just a food for thought. As aggregations are applied on raw data and we are not handling nested aggregations , I presume we don't want to allow duplicate aggregations. If previous aggregation results are not cached than duplicating same aggregation operation on huge record set can be costly. So, should we consider using Set Interface (Hashset)?

File: .gitignore [hunk 4]
```diff
@@ -3,4 +3,5 @@ target/
 *.iml
 __pycache__/
 
-.flattened-pom.xml
\ No newline at end of file
+.flattened-pom.xml
+*.patch
```
Comment (kolea2) [inline .gitignore:7]: was this change intentional?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/GqlQuery.java [hunk 5]
```diff
@@ -494,26 +511,17 @@ public boolean equals(Object obj) {
...
 
   @Override
-  void populatePb(com.google.datastore.v1.RunQueryRequest.Builder requestPb) {
+  public void populatePb(com.google.datastore.v1.RunQueryRequest.Builder requestPb) {
     requestPb.setGqlQuery(toPb());
   }
 
   @Override
-  Query<V> nextQuery(com.google.datastore.v1.RunQueryResponse responsePb) {
+  public RecordQuery<V> nextQuery(com.google.datastore.v1.RunQueryResponse responsePb) {
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/GqlQuery.java:524]: same here

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/RecordQuery.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.datastore;
+
+import com.google.api.core.InternalApi;
+import com.google.cloud.datastore.Query.ResultType;
+
+@InternalApi
+public interface RecordQuery<V>{
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/RecordQuery.java:22]: is this still needed?

File: google-cloud-datastore/clirr-ignored-differences.xml [hunk 7]
```diff
@@ -11,4 +11,9 @@
         <method>com.google.datastore.v1.ReserveIdsResponse reserveIds(com.google.datastore.v1.ReserveIdsRequest)</method>
         <differenceType>7012</differenceType>
     </difference>
+    <difference>
```
Comment (kolea2) [inline google-cloud-datastore/clirr-ignored-differences.xml:14]: as we are now on Java 8, please opt for a default method on the interface and remove this exception

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/AggregationBuilder.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.datastore.aggregation;
+
+public interface AggregationBuilder<A extends Aggregation> {
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/AggregationBuilder.java:19]: most builders in the library are simply static classes - do we need this level of abstraction?

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/AggregationBuilder.java (hunk 8 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/AggregationBuilder.java:19]: Hi @kolea2 , It will be required while adding aggregation to a query.

`
Query<AggregationResult> aggregationQuery = Query.newAggregationQueryBuilder()
  .addAggregation(count().as("total"))
  .addAggregation(sum("cost").as("total_cost"))
  .over(query)
  .build();
`

In this case, both sum and count are returning the instances of AggregationBuilder, and aggregationQuery can use the builders without being aware of the concrete builder implementation.

If we don't do this, then we need to have `N `overloaded version of `addAggregation` method for `N` type of aggregation query we will be supporting in future.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+  public GqlQuery<?> getNestedGqlQuery() {
+    return nestedGqlQuery;
+  }
+
+  public Mode getMode() {
+    return mode;
+  }
+
+  public static class Builder {
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java:67]: Hi @gforgurups, In a way we can do that. But we need AggregationQuery's `Builder` due to following reasons:

- End user is accustomed to using builder (ex: Query.newEntityQueryBuilder, Query.newKeyQueryBuilder etc.), and we created a new `Query.newAggregationQueryBuilder` to access this builder class. If we expose AggregationQuery class directly we will be breaking the pattern established in their mind.
- If we see from code perspective, we have followed the existing convention and maintain the consistency with other Query classes by exposing this builder.

These two reasons are good enough for us to keep having this Builder, let me know if you think otherwise.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/CountAggregation.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+import com.google.protobuf.Int64Value;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+public class CountAggregation extends Aggregation {
+
+  private final long limit;
+
+  private CountAggregation(String alias, long limit) {
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/CountAggregation.java:30]: that make sense @gforgurups . There is nothing wrong in giving creational access to Aggregations. I'll make this change and add one more version `addAggregation(Aggregation)`method along with `addAggregation(AggregationBuilder)`. so that user can pass Aggregation directly without it's builder.

[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/aggregation/CountAggregation.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/5097cd69fc6c1684103fa4500c289b86fe81d1ce.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.api.core.InternalApi;
+import com.google.cloud.Timestamp;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+public class AggregationResults implements Iterable<AggregationResult> {
+
+  private final List<AggregationResult> aggregationResults;
+  private Timestamp readTime;
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java]: Yes, fixed in the latest version of code.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQueryProtoPreparer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQueryProtoPreparer.java:27]: Hi @gforgurups 

For `query` parameter, we have already done the [validation here](https://github.com/googleapis/java-datastore/pull/823/files#diff-4d491d7fe61ca45f3587ec96ad5c619b7ec0c147f33b150b47d681195f814536R108-R109) at the creation time itself. At this point in time when we. are populating protos we are certain that it wont' be null.


and regarding `com.google.datastore.v1.Query.newBuilder()` is a generated code from our Proto definitions. Hence it is considered safe enough to not to return null.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java (hunk 3 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/6508ff6dc97171ee652a8c9b189f7425e9d4bd45.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java (hunk 1 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java:31]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/6508ff6dc97171ee652a8c9b189f7425e9d4bd45.

File: .gitignore (hunk 4 above)
Comment (jainsahab) [reply .gitignore:7]: I was maintaining multiple patches in my local, so thought it would be useful. Do you want me to remove this ?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/RecordQuery.java (hunk 6 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/RecordQuery.java:22]: Hi @kolea2 , before this feature, `Query` (top level class) was used heavily in the code to generalise between GqlQuery and StructuredQuery,  We used `Query` reference so that different components can still remain oblivious to the underlying Query. 

With the arrival of AggregationQuery, we had to move entity records specific behaviour to `RecordQuery` to keep the `Query` clean enough to be the top level class. Instead of using `Query` reference, existing query handling mechanism uses `RecordQuery` so that existing code doesn't change much.

If we dont do this, then we need to write code to handle StructuredQuery and GqlQuery separately.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/GqlQuery.java [hunk 12]
```diff
@@ -494,26 +511,17 @@ public boolean equals(Object obj) {
...
-      queryPb.addPositionalBindings(argument.toPb());
-    }
-    return queryPb.build();
+    GqlQueryProtoPreparer protoPreparer = new GqlQueryProtoPreparer();
+    return protoPreparer.prepare(this);
   }
 
   @Override
-  void populatePb(com.google.datastore.v1.RunQueryRequest.Builder requestPb) {
+  public void populatePb(com.google.datastore.v1.RunQueryRequest.Builder requestPb) {
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/GqlQuery.java:521]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/5f1e419a3e30aca687e49eadccd0a135a8f1b2e0.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/GqlQuery.java (hunk 5 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/GqlQuery.java:527]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/5f1e419a3e30aca687e49eadccd0a135a8f1b2e0.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+import static com.google.cloud.datastore.AggregationQuery.Mode.STRUCTURED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.cloud.datastore.aggregation.Aggregation;
+import com.google.cloud.datastore.aggregation.AggregationBuilder;
+import java.util.ArrayList;
+import java.util.List;
+
+public class AggregationQuery extends Query<AggregationResults> {
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java:76]: ✅  Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/5e2fcb3d4aeee19181d50d6b64b4d54d7363a95b.

File: google-cloud-datastore/clirr-ignored-differences.xml (hunk 7 above)
Comment (kolea2) [reply google-cloud-datastore/clirr-ignored-differences.xml:14]: I am still in favor of adding a default method rather than adding a breaking change here - these other exceptions (I believe) were when we were still on Java 7.

Comment (jainsahab) [reply google-cloud-datastore/clirr-ignored-differences.xml:14]: Just being the devil's advocate here and contradicting the `clirr` check. How is adding a  new method can be considered a breaking change? We are not breaking any existing expectations of an end user here, Just providing them with more capabilities than the older versions.

Comment (jainsahab) [reply google-cloud-datastore/clirr-ignored-differences.xml:14]: BTW, i quickly added a dummy default method in the above mentioned interface, and clirr still failed.
image

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationQueryTest.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+      .build();
+
+  @Rule
+  public ExpectedException exceptionRule = ExpectedException.none();
+
+  @Test
+  public void testAggregations() {
+    AggregationQuery aggregationQuery = Query.newAggregationQueryBuilder()
+        .setNamespace(NAMESPACE)
+        .addAggregation(new CountAggregation("total_upto_100", 100))
```
Comment (vishwarajanand) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationQueryTest.java:52]: The [design doc](http://go/java-datastore-count-design#heading=h.nxddek4w5dli) mentions this:

```
  .addAggregation(count().as("total"))           // not CountAggregation
```

Comment (jainsahab) [reply google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationQueryTest.java:52]: Yes, both syntax are supported.

https://github.com/googleapis/java-datastore/pull/823#discussion_r971588252

The next test in the same file follows the builder construct.

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationResultTest.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+public class AggregationResultTest {
+
+  @Test
+  public void shouldGetAggregationResultValueByAlias() {
+    AggregationResult aggregationResult = new AggregationResult(ImmutableMap.of(
+        "count", LongValue.of(45),
+        "count_upto_30", LongValue.of(30)
+    ));
+
+    assertThat(aggregationResult.get("count"), equalTo(45L));
```
Comment (kolea2) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationResultTest.java:33]: I would use `Truth.assertThat...` to be consistent with other tests.

**Review by kolea2 — CHANGES_REQUESTED** [review]
looking really good!

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 16]
```diff
@@ -337,33 +360,11 @@ public Iterator<Entity> get(Iterable<Key> keys, ReadOption... options) {
...
-      if (optionsByType.containsKey(ReadTime.class)) {
-        readOptionsPb =
-            com.google.datastore.v1.ReadOptions.newBuilder()
-                .setReadTime(((ReadTime) optionsByType.get(ReadTime.class)).time().toProto())
-                .build();
-      }
+      return this.readOptionProtoPreparer.prepare(Arrays.asList(options));
     }
-    return readOptionsPb;
+    return null;
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:367]: Hi @kolea2 , I've maintained the existing contract of this method, that is to return null when no `ReadOptions` are provided by the user.
Since  we are using Java 8, how about if we use `Optional` here ? seems like a perfect use case.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  AggregationQuery(String namespace, GqlQuery<?> gqlQuery) {
+    super(namespace);
+    this.nestedGqlQuery = gqlQuery;
+    this.mode = GQL;
+  }
+
+  /** Returns the {@link Aggregation}(s) for this Query. */
+  public Set<Aggregation> getAggregations() {
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java:98]: https://github.com/googleapis/java-datastore/pull/823/#discussion_r972614041

Few things like multiple aggregations with same aliases can be solved by using set, and if we pass duplicate aggregation to backend it'll throw errors like:
image
By using a`Set` we can handle this mistake by the user and return the results gracefully.

Proto defines this field with `repeated` keyword. Though order doesn't matter in this case, As the enduser  gets the map (`Map<Alias, AggregationValue>`) in result. If we want to preserve the order of aggregations in aggregation query, we can use `LinkedHashSet`. Let me know what you think.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/Query.java [hunk 18]
```diff
@@ -156,26 +156,18 @@ static ResultType<?> fromPb(com.google.datastore.v1.EntityResult.ResultType type
...
-    return resultType;
-  }
-
   public String getNamespace() {
     return namespace;
   }
 
   ToStringHelper toStringHelper() {
-    return MoreObjects.toStringHelper(this).add("type", resultType).add("namespace", namespace);
+    return MoreObjects.toStringHelper(this).add("namespace", namespace);
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/Query.java:168]: `type` here represents ENTITY, PROJECTIONS or KEY. we have removed the coupling of such things from `Query`  and made it generic. Hence, this is removed from here and moved down in the hierarchy. ex: `StructuredQuery#toString()`

File: .gitignore (hunk 4 above)
Comment (jainsahab) [reply .gitignore]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/8c81b9f201102b4db79d195263d15fb06e4985d3.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/Datastore.java [hunk 19]
```diff
@@ -461,4 +461,54 @@ public interface Datastore extends Service<DatastoreOptions>, DatastoreReaderWri
...
+
+  /**
+   * Submits a {@link AggregationQuery} and returns {@link AggregationResults}.
+   * {@link ReadOption}s can be specified if desired.
+   *
+   * <p>Example of running an {@link AggregationQuery} to find the count of entities of one kind.
+   *
+   * <p>{@link StructuredQuery} example:</p>
+   * <pre>{@code
+   * import static com.google.cloud.datastore.aggregation.Aggregation.count;
```
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/Datastore.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/8c81b9f201102b4db79d195263d15fb06e4985d3.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Replaced the code dealing with null here with `Optional` here https://github.com/googleapis/java-datastore/pull/823/commits/1584d9f88ae7af6a87223daaf5d0af6690e5a017.

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationResultTest.java (hunk 15 above)
Comment (jainsahab) [reply google-cloud-datastore/src/test/java/com/google/cloud/datastore/AggregationResultTest.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/bc425ddc2a887020beb6fd49355ee96ac94a76b4.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: great!

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationQuery.java:96]: got it - as long as it's doing what the backend expects, sounds good to me!

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * The result of an {@link AggregationQuery} query submission. Contains a {@link
+ * List<AggregationResult>} and readTime {@link Timestamp} in it.
+ *
+ * <p>This can be used to iterate over underlying {@link List<AggregationResult>} directly.
+ *
+ * <p>Though {@link com.google.cloud.datastore.aggregation.CountAggregation} is guaranteed to return
+ * only one {@link AggregationResult} as part of its execution.
+ *
+ * <p>In the future, we might support more complex {@link AggregationQuery} that might result in
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java:36]: nit - I would remove this to keep the docs just about the class functionality itself

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/RetryAndTraceDatastoreRpcDecorator.java [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Status;
+import java.util.concurrent.Callable;
+
+/**
+ * An implementation of {@link DatastoreRpc} which acts as a Decorator and decorates the underlying
+ * {@link DatastoreRpc} with the logic of retry and Traceability.
+ */
+@InternalApi
+public class RetryAndTraceDatastoreRpcDecorator implements DatastoreRpc {
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/RetryAndTraceDatastoreRpcDecorator.java:52]: because this is only implemented for aggregation queries, perhaps this name should be something different?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQuery.java [hunk 22]
```diff
@@ -1018,6 +1019,8 @@ public Integer getLimit() {
 
   public abstract Builder<V> toBuilder();
 
+  @InternalApi
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/StructuredQuery.java:1022]: I don't think we should change this to internal

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java [hunk 23]
```diff
@@ -16,6 +16,8 @@
 
 package com.google.cloud.datastore;
 
+import static com.google.cloud.datastore.ReadOption.transactionId;
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java:19]: I would be inclined to revert this, as it makes the readability more confusing (`transactionId(transactionId)`)

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/execution/response/AggregationQueryResponseTransformer.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    implements ResponseTransformer<RunAggregationQueryResponse, AggregationResults> {
+
+  @Override
+  public AggregationResults transform(RunAggregationQueryResponse response) {
+    Timestamp readTime = Timestamp.fromProto(response.getBatch().getReadTime());
+    List<AggregationResult> aggregationResults =
+        response.getBatch().getAggregationResultsList().stream()
+            .map(
+                aggregationResult -> new AggregationResult(resultWithLongValues(aggregationResult)))
+            .collect(Collectors.toCollection(LinkedList::new));
```
Comment (kolea2) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/execution/response/AggregationQueryResponseTransformer.java:44]: why a linked list here?

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/ProtoTestData.java [hunk 25]
```diff
@@ -63,6 +66,17 @@ public static PropertyReference propertyReference(String value) {
...
   }
 
+  public static Aggregation countAggregation(String alias) {
+    return Aggregation.newBuilder().setAlias(alias).setCount(Count.newBuilder().build()).build();
+  }
+
+  public static Aggregation countAggregation(String alias, long limit) {
+    return Aggregation.newBuilder()
+        .setAlias(alias)
+        .setCount(Count.newBuilder().setUpTo(Int64Value.of(limit)))
```
Comment (kolea2) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/ProtoTestData.java:76]: I don't think we want to use up to? Rather, can we use limit on the underlying query?

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/execution/AggregationQueryExecutorTest.java [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+                    new AggregationResult(
+                        ImmutableMap.of(
+                            "count", LongValue.of(209), "property_2", LongValue.of(100))),
+                    new AggregationResult(
+                        ImmutableMap.of(
+                            "count", LongValue.of(509), "property_2", LongValue.of(100)))),
+                Timestamp.fromProto(runAggregationQueryResponse.getBatch().getReadTime())));
+  }
+
+  private RunAggregationQueryResponse dummyAggregationQueryResponse() {
```
Comment (kolea2) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/execution/AggregationQueryExecutorTest.java:146]: rename to `placeholderAggregationQueryResponse`

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreTest.java [hunk 27]
```diff
@@ -1311,4 +1344,28 @@ public void testQueryWithStartCursor() {
     assertEquals(cursor2, cursor1);
     datastore.delete(entity1.getKey(), entity2.getKey(), entity3.getKey());
   }
+
+  private RunAggregationQueryResponse dummyAggregationQueryResponse() {
```
Comment (kolea2) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreTest.java:1348]: rename to `placeholderAggregationQueryResponse`

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java (hunk 23 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java:19]: how about we change the method name to `withTransaction`? In that case it becomes `withTransaction(transactionId)`.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/execution/response/AggregationQueryResponseTransformer.java (hunk 24 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/execution/response/AggregationQueryResponseTransformer.java:44]: while writing this i felt LinkedList will be more performant here because:
 - If we have too many `AggregationResult` coming in, then `ArrayList` will need to double it's capacity everytime it's full and copy the existing data in it whereas in `LinkedList` it'll be a O(1) operation.
 - We are exposing the result as an iterator anyway and won't need a case where we need to access the result using random index where `ArrayList` is optimized.

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreTest.java (hunk 27 above)
Comment (jainsahab) [reply google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreTest.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/a5b462332db2f8b88db4016d2f83046c89e74acd.

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/execution/AggregationQueryExecutorTest.java (hunk 26 above)
Comment (jainsahab) [reply google-cloud-datastore/src/test/java/com/google/cloud/datastore/execution/AggregationQueryExecutorTest.java]: Fixed as part of https://github.com/googleapis/java-datastore/pull/823/commits/a5b462332db2f8b88db4016d2f83046c89e74acd.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java (hunk 20 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/AggregationResults.java]: Fixed as part of https://github.com/googleapis/java-datastore/commit/a5b462332db2f8b88db4016d2f83046c89e74acd.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java (hunk 23 above)
Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java:19]: that sounds good!

Comment (kolea2) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java:19]: or `setTransactionId` also sgtm

**Review by gforgurups — APPROVED** [review]

**Review by kolea2 — APPROVED** [review]

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java (hunk 23 above)
Comment (jainsahab) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java:19]: addressed in https://github.com/googleapis/java-datastore/pull/1041.
